File: algorithms-python/heap/IndexMinHeap.py
class IndexMinHeap:
    def __init__(self, N):
        self.pq=[] # position of element on heap
        self.N=N
        self.qp=[-1 for x in  range(self.N)]
        self.keys=[]



    def swim(self,n):
        while n!=1 and self.less(n,n//2):
            self.exch(n,n//2)
            n=n//2

    def sink(self,n):
        k=0
        while 2*n<self.N:
            if self.less(2*n,2*n + 1):
                k=n
            else:
                k=n+1
            if self.less(n,k):
                self.exch(n,k)
                n=k
            else:
                break



    def exch(self,i,j):
        self.pq[i],self.pq[j] = self.pq[j], self.pq[i]
        self.qp[self.pq[i]]=i
        self.qp[self.pq[j]]=j

    def less(self,i,j):
        if self.keys[self.pq[i]]< self.keys[self.pq[j]]:
            True
        else:
            False

    def insert(self,i,key):
         self.N+=1
         self.pq[self.N]=i
         self.keys[self.N]=key
         self.qp[i]=self.N
         self.swim(self.N)

    def delete(self,i,key):
        heapIndex=self.qp[i];
        self.exch(1,heapIndex)
        self. N-=1
        self.sink(self.N)

    def changekey(self,i,key):
        heapIndex=self.qp[i]
        self.keys[i] =key
        #can be greater or smaller
        self.swim(heapIndex)
        self.sink(heapIndex)

    def decreaseKey(self,i,key):
        heapIndex=self.pq[i]
        self.keys[i]=key;
        self.swim(heapIndex)


    def contains(self,i):
        if i <1 or i>self.N:
            raise ValueError("Index out of bound")
        else:
            self.qp[i]!=-1

File: Algorithms/src/heap/IndexMinHeap.java
package heap;

import java.util.NoSuchElementException;

/**
 * Created by vasantbedadala on 2/15/16.
 */
public class IndexMinHeap<T extends Comparable<T>> {

    //heap positions
    int [] pq;
    //array of keys to heap indexes
    int [] qp;
    Object [] keys;


    int size;
    int N;

    public IndexMinHeap(int size) {
        pq = new int[size + 1];
        qp = new int[size + 1];
        keys = new Object[size +1];
        this.size = size;
        this.N=0;
    }


    public T min() {
        if(!isEmpty()) {
            return(T) keys[pq[1]];
        }
        return null;
    }

    public int minIndex() {
        if(!isEmpty()) {
            //since heap is one indexed
            return pq[1];
        }
        return -1;
    }

    public T delMin() {
        if(!isEmpty()) {
            T min = (T) keys[pq[1]];
            exch( 1, N--);
            qp[pq[1]]=-1;
            sink(1);
                return min;
        }
        return null;
    }

    public T getKey(int i) {
        return (T) keys[i];
    }
    public void changeKey(int i, T key) {
        keys[i]=key;
        swim(qp[i]);
        sink(qp[i]);
    }

    public void decreaseKey(int i, int key) {
       if(!containsIndex(i)) {
           throw new NoSuchElementException("Index not in the priority queue");

       }
        keys[i]=key;    
        swim(qp[i]);
    }

    public void increaseKey(int i, int key) {
        if(!containsIndex(i)) {
            throw new NoSuchElementException("Index not in the priority queue");

        }
        keys[i]=key;
        sink(qp[i]);
    }


    private boolean isMinHeap(int k) {
        int left = 2 * k;
        int right = 2 * k + 1;
        if (((T)keys[pq[k]]).compareTo((T)keys[left])>=0) return false;
        if (((T)keys[pq[k]]).compareTo((T)keys[right])>=0) return false;
        return isMinHeap(left) && isMinHeap(right);
    }

    private void sink(int key) {
        while (2 * key <=N) {
            int k = 2 * key;
            if (k<N && ((T)keys[pq[k]]).compareTo((T)keys[pq[k + 1]]) >=1) {
                k = k + 1;
            }
            if (((T)keys[pq[key]]).compareTo((T) keys[pq[k]]) <0)
                break;
            exch(key, k);
            key = k;
        }
    }


    private void swim(int key) {
        //while not at root and less than parent
        while (key != 1 && (((T)keys[pq[key]]).compareTo((T)keys[pq[(int) Math.floor(key / 2)]])<0)) {
            exch(key, (int) Math.floor(key / 2));
            key = (int) Math.floor(key / 2);
        }
    }

    private void exch( int i, int j) {
        int temp = pq[i];
        pq[i] = pq[j];
        pq[j] = temp;
        qp[pq[i]]=i;
        qp[pq[j]]=j;

    }



    public void insert(int i,T key) {

        this.N++;
        pq[N] = i;
        qp[i] =N;
        keys[i]=key;
        swim(N);
    }

    // helper function to double the size of the heap array
    private void resize(int capacity) {
        int[] temp = new int[capacity];
        for (int i = 1; i <= size; i++) {
            temp[i] = pq[i];
        }
        pq = temp;
    }

    public boolean isEmpty() {
        return this.N==0;
    }

    public int size() {
        return this.N;
    }

    public boolean containsIndex(int index) {
        return qp[index]!=-1;
    }
}
